File: solutions/php/say/1/Say.php
<?php

/*
 * By adding type hints and enabling strict type checking, code can become
 * easier to read, self-documenting and reduce the number of potential bugs.
 * By default, type declarations are non-strict, which means they will attempt
 * to change the original type to match the type specified by the
 * type-declaration.
 *
 * In other words, if you pass a string to a function requiring a float,
 * it will attempt to convert the string value to a float.
 *
 * To enable strict mode, a single declare directive must be placed at the top
 * of the file.
 * This means that the strictness of typing is configured on a per-file basis.
 * This directive not only affects the type declarations of parameters, but also
 * a function's return type.
 *
 * For more info review the Concept on strict type checking in the PHP track
 * <link>.
 *
 * To disable strict typing, comment out the directive below.
 */

declare(strict_types=1);

function say(int $number): string
{
    if ($number < 0 || $number >= 1000000000000) {
        throw new InvalidArgumentException("Input out of range");
    }

    if ($number === 0) {
        return "zero";
    }

    $scales = [
        ['', ''],
        ['thousand', 'thousand'],
        ['million', 'million'],
        ['billion', 'billion'],
    ];

    $chunks = [];
    $scaleIndex = 0;

    while ($number > 0) {
        $chunkValue = $number % 1000;
        $number = intdiv($number, 1000);

        if ($chunkValue > 0) {
            $chunkWords = convertChunk($chunkValue);
            if ($scaleIndex > 0) {
                $chunkWords .= ' ' . $scales[$scaleIndex][0];
            }
            $chunks[] = $chunkWords;
        }

        $scaleIndex++;
    }

    return implode(' ', array_reverse($chunks));
}

function convertChunk(int $number): string
{
    $ones = [
        '', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine',
        'ten', 'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen'
    ];

    $tens = [
        '', '', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety'
    ];

    $result = '';

    // Handle hundreds
    if ($number >= 100) {
        $result .= $ones[intdiv($number, 100)] . ' hundred';
        $number %= 100;
        if ($number > 0) {
            $result .= ' ';
        }
    }

    // Handle tens and ones
    if ($number > 0) {
        if ($number < 20) {
            $result .= $ones[$number];
        } else {
            $tensDigit = intdiv($number, 10);
            $onesDigit = $number % 10;
            
            $result .= $tens[$tensDigit];
            
            if ($onesDigit > 0) {
                $result .= '-' . $ones[$onesDigit];
            }
        }
    }

    return $result;
}

File: solutions/java/say/1/src/main/java/Say.java
public class Say {
    
    private static final String[] ONES = {
        "", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
        "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen",
        "seventeen", "eighteen", "nineteen"
    };
    
    private static final String[] TENS = {
        "", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"
    };
    
    private static final String[] SCALES = {"", "thousand", "million", "billion"};
    
    public String say(long number) {
        // Validate input range
        if (number < 0 || number > 999_999_999_999L) {
            throw new IllegalArgumentException("Number must be between 0 and 999,999,999,999.");
        }
        
        // Special case for zero
        if (number == 0) {
            return "zero";
        }
        
        StringBuilder result = new StringBuilder();
        int scaleIndex = 0;
        
        // Process number in chunks of 3 digits (ones, thousands, millions, billions)
        while (number > 0) {
            int chunk = (int) (number % 1000);
            if (chunk != 0) {
                String chunkWords = convertChunk(chunk);
                if (scaleIndex > 0) {
                    chunkWords += " " + SCALES[scaleIndex];
                }
                if (result.length() > 0) {
                    result.insert(0, chunkWords + " ");
                } else {
                    result.append(chunkWords);
                }
            }
            number /= 1000;
            scaleIndex++;
        }
        
        return result.toString();
    }
    
    // Convert a number 1-999 to words (no scale suffix)
    private String convertChunk(int num) {
        StringBuilder chunk = new StringBuilder();
        
        // Handle hundreds place
        if (num >= 100) {
            chunk.append(ONES[num / 100]).append(" hundred");
            num %= 100;
            if (num > 0) {
                chunk.append(" ");
            }
        }
        
        // Handle tens and ones place
        if (num >= 20) {
            chunk.append(TENS[num / 10]);
            if (num % 10 > 0) {
                chunk.append("-").append(ONES[num % 10]);
            }
        } else if (num > 0) {
            chunk.append(ONES[num]);
        }
        
        return chunk.toString();
    }
}